### ISSUE TITLE
Crash on iOS Smartlook SDK 1.8.3

### ISSUE THREAD

**Issue by mithilesh-parmar** [smartlook/smartlook-mobile-issue-tracker#123]
**Platform**
iOS

**Technology**
Swift

**Smartlook SDK version**
1.8.3

**Summary**

**Steps to reproduce [optional]**
Not able to reproduce, happening randomly. 

**Relevant logs and/or screenshots**
```
OS Version: iOS 12.5.5 (16H62)
Report Version: 104

Exception Type: EXC_BAD_ACCESS (SIGBUS)
Exception Codes: BUS_NOOP
Crashed Thread: 0

Application Specific Information:
Exception 1, Code 16, Subcode 8 >
Attempted to dereference garbage pointer 0x8000000000000010.

Thread 0 Crashed:
0   Smartlook                       0x10218f6a4         sl_swizzle_selector_macro_template
1   Smartlook                       0x10218fe24         sl_swizzle_selector_macro_template
2   Smartlook                       0x10218f960         sl_swizzle_selector_macro_template
3   QuartzCore                      0x3719abfec         CA::Display::DisplayLink::dispatch_items
4   IOKit                           0x36964f4ac         IODispatchCalloutFromCFMessage
5   CoreFoundation                  0x368fec98c         __CFMachPortPerform
6   CoreFoundation                  0x369013590         __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE1_PERFORM_FUNCTION__
7   CoreFoundation                  0x369012cdc         __CFRunLoopDoSource1
8   CoreFoundation                  0x36900db00         __CFRunLoopRun
9   CoreFoundation                  0x36900cfb0         CFRunLoopRunSpecific
10  GraphicsServices                0x36d4ac798         GSEventRunModal
11  UIKitCore                       0x3c1c8ec34         UIApplicationMain
12  libswiftUIKit.dylib             0x3c537b470         UIApplicationMain
13  Powerplay                       0x2009427d4         <redacted>
14  libdyld.dylib                   0x3686378dc         start

Thread 0 Crashed:
0   Smartlook                       0x10218f6a4         sl_swizzle_selector_macro_template
1   Smartlook                       0x10218fe24         sl_swizzle_selector_macro_template
2   Smartlook                       0x10218f960         sl_swizzle_selector_macro_template
3   QuartzCore                      0x3719abfec         CA::Display::DisplayLink::dispatch_items
4   IOKit                           0x36964f4ac         IODispatchCalloutFromCFMessage
5   CoreFoundation                  0x368fec98c         __CFMachPortPerform
6   CoreFoundation                  0x369013590         __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE1_PERFORM_FUNCTION__
7   CoreFoundation                  0x369012cdc         __CFRunLoopDoSource1
8   CoreFoundation                  0x36900db00         __CFRunLoopRun
9   CoreFoundation                  0x36900cfb0         CFRunLoopRunSpecific
10  GraphicsServices                0x36d4ac798         GSEventRunModal
11  UIKitCore                       0x3c1c8ec34         UIApplicationMain
12  libswiftUIKit.dylib             0x3c537b470         UIApplicationMain
13  Powerplay                       0x2009427d4         <redacted>
14  libdyld.dylib                   0x3686378dc         start

Thread 1
0   libsystem_kernel.dylib          0x3688afb74         __workq_kernreturn
1   libsystem_pthread.dylib         0x3689cc134         _pthread_wqthread

Thread 2
0   libsystem_pthread.dylib         0x3689cecd0         start_wqthread

Thread 3
0   libsystem_kernel.dylib          0x3688afb74         __workq_kernreturn
1   libsystem_pthread.dylib         0x3689cc134         _pthread_wqthread

Thread 4
0   libswiftCore.dylib              0x3c48904a4         _HashTable.previousHole
1   Smartlook                       0x1021ea1f8         sl_swizzle_selector_macro_template
2   Smartlook                       0x1021e8f84         sl_swizzle_selector_macro_template
3   Smartlook                       0x1021e93ac         sl_swizzle_selector_macro_template
4   Smartlook                       0x10219395c         sl_swizzle_selector_macro_template
5   Smartlook                       0x102193d08         sl_swizzle_selector_macro_template
6   Smartlook                       0x1021d5418         sl_swizzle_selector_macro_template
7   Smartlook                       0x10219adb4         sl_swizzle_selector_macro_template
8   Smartlook                       0x1021aa994         sl_swizzle_selector_macro_template
9   Smartlook                       0x1021aa9b4         sl_swizzle_selector_macro_template
10  CoreData                        0x36e7852ac         developerSubmittedBlockToNSManagedObjectContextPerform
11  libdispatch.dylib               0x3685b67d0         _dispatch_client_callout
12  libdispatch.dylib               0x368564c18         _dispatch_lane_barrier_sync_invoke_and_complete
13  CoreData                        0x36e64f784         -[NSManagedObjectContext performBlockAndWait:]
14  Smartlook                       0x10219acd4         sl_swizzle_selector_macro_template
15  Smartlook                       0x10213382c         sl_swizzle_selector_macro_template
16  Smartlook                       0x1021338ac         sl_swizzle_selector_macro_template
17  Foundation                      0x36a554a74         __NSOQSchedule_f
18  libdispatch.dylib               0x3685b5a34         _dispatch_call_block_and_release
19  libdispatch.dylib               0x3685b67d0         _dispatch_client_callout
20  libdispatch.dylib               0x36855b018         _dispatch_continuation_pop$VARIANT$mp
21  libdispatch.dylib               0x36855a6dc         _dispatch_async_redirect_invoke
22  libdispatch.dylib               0x36856702c         _dispatch_root_queue_drain
23  libdispatch.dylib               0x3685678d0         _dispatch_worker_thread2
24  libsystem_pthread.dylib         0x3689cc1b0         _pthread_wqthread

Thread 5
0   libsystem_kernel.dylib          0x3688b0408         fsync
1   Foundation                      0x36a559cd4         _NSWriteDataToFileWithExtendedAttributes
2   GoogleUtilities                 0x101959370         <redacted>
3   GoogleUtilities                 0x1019592b8         <redacted>
4   Foundation                      0x36a4a98dc         -[NSFileCoordinator _invokeAccessor:thenCompletionHandler:]
5   Foundation                      0x36a630aa4         __98-[NSFileCoordinator coordinateReadingItemAtURL:options:writingItemAtURL:options:error:byAccessor:]_block_invoke
6   Foundation                      0x36a631f8c         __110-[NSFileCoordinator _coordinateReadingItemAtURL:options:writingItemAtURL:options:error:byAccessor:]_block_invoke.359
7   Foundation                      0x36a4a9b0c         -[NSFileCoordinator(NSPrivate) _withAccessArbiter:invokeAccessor:orDont:andRelinquishAccessClaim:]
8   Foundation                      0x36a4b2d88         -[NSFileCoordinator(NSPrivate) _coordinateReadingItemAtURL:options:writingItemAtURL:options:error:byAccessor:]
9   Foundation                      0x36a4b2978         -[NSFileCoordinator coordinateReadingItemAtURL:options:writingItemAtURL:options:error:byAccessor:]
10  GoogleUtilities                 0x1019591a0         <redacted>
11  FirebaseCoreDiagnostics         0x10160260c         FIRPopulateProtoWithInfoPlistValues
12  FirebaseCoreDiagnostics         0x1016023d8         FIRPopulateProtoWithInfoPlistValues
13  libdispatch.dylib               0x3685b5a34         _dispatch_call_block_and_release
14  libdispatch.dylib               0x3685b67d0         _dispatch_client_callout
15  libdispatch.dylib               0x36855f320         _dispatch_lane_serial_drain$VARIANT$mp
16  libdispatch.dylib               0x36855fe3c         _dispatch_lane_invoke$VARIANT$mp
17  libdispatch.dylib               0x3685684a8         _dispatch_workloop_worker_thread
18  libsystem_pthread.dylib         0x3689cc110         _pthread_wqthread

Thread 6 name: com.apple.uikit.eventfetch-thread
0   libsystem_kernel.dylib          0x3688a40f4         mach_msg_trap
1   libsystem_kernel.dylib          0x3688a359c         mach_msg
2   CoreFoundation                  0x369012910         __CFRunLoopServiceMachPort
3   CoreFoundation                  0x36900d820         __CFRunLoopRun
4   CoreFoundation                  0x36900cfb0         CFRunLoopRunSpecific
5   Foundation                      0x36a447958         -[NSRunLoop(NSRunLoop) runMode:beforeDate:]
6   Foundation                      0x36a4477e8         -[NSRunLoop(NSRunLoop) runUntilDate:]
7   UIKitCore                       0x3c1d74750         -[UIEventFetcher threadMain]
8   Foundation                      0x36a57449c         __NSThread__start__
9   libsystem_pthread.dylib         0x3689cb2bc         _pthread_body
10  libsystem_pthread.dylib         0x3689cb21c         _pthread_start

Thread 7
0   libsystem_kernel.dylib          0x3688afb74         __workq_kernreturn
1   libsystem_pthread.dylib         0x3689cc134         _pthread_wqthread

Thread 8
0   libsystem_kernel.dylib          0x3688afb74         __workq_kernreturn
1   libsystem_pthread.dylib         0x3689cc134         _pthread_wqthread

Thread 9
0   libsystem_kernel.dylib          0x3688afb74         __workq_kernreturn
1   libsystem_pthread.dylib         0x3689cc134         _pthread_wqthread

Thread 10 name: JavaScriptCore bmalloc scavenger
0   libsystem_kernel.dylib          0x3688aeee4         __psynch_cvwait
1   libsystem_pthread.dylib         0x3689c3cf4         _pthread_cond_wait$VARIANT$mp
2   libc++.1.dylib                  0x36747608c         std::__1::condition_variable::wait
3   JavaScriptCore                  0x377560a9c         std::__1::condition_variable_any::wait<T>
4   JavaScriptCore                  0x377564a90         bmalloc::Scavenger::threadRunLoop
5   JavaScriptCore                  0x377564208         bmalloc::Scavenger::threadEntryPoint
6   JavaScriptCore                  0x377565c48         std::__1::__thread_proxy<T>
7   libsystem_pthread.dylib         0x3689cb2bc         _pthread_body
8   libsystem_pthread.dylib         0x3689cb21c         _pthread_start

Thread 11 name: com.apple.NSURLConnectionLoader
0   libsystem_kernel.dylib          0x3688a40f4         mach_msg_trap
1   libsystem_kernel.dylib          0x3688a359c         mach_msg
2   CoreFoundation                  0x369012910         __CFRunLoopServiceMachPort
3   CoreFoundation                  0x36900d820         __CFRunLoopRun
4   CoreFoundation                  0x36900cfb0         CFRunLoopRunSpecific
5   CFNetwork                       0x369ce5748         -[__CoreSchedulingSetRunnable runForever]
6   Foundation                      0x36a57449c         __NSThread__start__
7   libsystem_pthread.dylib         0x3689cb2bc         _pthread_body
8   libsystem_pthread.dylib         0x3689cb21c         _pthread_start

Thread 12
0   libsystem_kernel.dylib          0x3688b0408         fsync
1   libsqlite3.dylib                0x3699f43e8         sqlite3_free_table
2   libsqlite3.dylib                0x3699fee9c         sqlite3_randomness
3   libsqlite3.dylib                0x3699f3cf4         sqlite3_free_table
4   libsqlite3.dylib                0x3699de150         sqlite3_value_text
5   libsqlite3.dylib                0x3699a9df0         sqlite3_finalize
6   libsqlite3.dylib                0x3699d6afc         sqlite3_step
7   libsqlite3.dylib                0x3699c6b3c         sqlite3_step
8   FirebaseRemoteConfig            0x1016ed99c         <redacted>
9   FirebaseRemoteConfig            0x1016ed754         <redacted>
10  libdispatch.dylib               0x3685b5a34         _dispatch_call_block_and_release
11  libdispatch.dylib               0x3685b67d0         _dispatch_client_callout
12  libdispatch.dylib               0x36855f320         _dispatch_lane_serial_drain$VARIANT$mp
13  libdispatch.dylib               0x36855fe3c         _dispatch_lane_invoke$VARIANT$mp
14  libdispatch.dylib               0x3685684a8         _dispatch_workloop_worker_thread
15  libsystem_pthread.dylib         0x3689cc110         _pthread_wqthread

Thread 13
0   libnetwork.dylib                0x36b137fd8         nw_protocol_definition_is_equal
1   libnetwork.dylib                0x36b16d224         nw_parameters_copy_tcp_options
2   libnetwork.dylib                0x36b16e6e4         nw_parameters_set_no_delay
3   CFNetwork                       0x369e9a2fc         TCPIOConnection::_applyPropertiesToConnection
4   CFNetwork                       0x369e95838         TCPIOConnection::_startConnection
5   CFNetwork                       0x369e08824         Tube::_onqueue_startConnection
6   CFNetwork                       0x369e0814c         Tube::_onqueue_prepConnection
7   CFNetwork                       0x369e087d0         ___ZN4Tube23_onqueue_prepConnectionEU13block_pointerFvvEU13block_pointerFviE_block_invoke_2.63
8   libdispatch.dylib               0x3685b5a34         _dispatch_call_block_and_release
9   libdispatch.dylib               0x3685b67d0         _dispatch_client_callout
10  libdispatch.dylib               0x36855f320         _dispatch_lane_serial_drain$VARIANT$mp
11  libdispatch.dylib               0x36855fe70         _dispatch_lane_invoke$VARIANT$mp
12  libdispatch.dylib               0x368561108         _dispatch_workloop_invoke$VARIANT$mp
13  libdispatch.dylib               0x3685684a8         _dispatch_workloop_worker_thread
14  libsystem_pthread.dylib         0x3689cc110         _pthread_wqthread

Thread 14
0   libsystem_kernel.dylib          0x3688af400         __semwait_signal
1   libsystem_c.dylib               0x368746568         nanosleep
2   libsystem_c.dylib               0x368746368         sleep
3   Sentry                          0x102069340         sentrycrashccd_init
4   libsystem_pthread.dylib         0x3689cb2bc         _pthread_body
5   libsystem_pthread.dylib         0x3689cb21c         _pthread_start

Thread 15 name: SentryCrash Exception Handler (Secondary)
0   libsystem_kernel.dylib          0x3688a40f4         mach_msg_trap
1   libsystem_kernel.dylib          0x3688a359c         mach_msg
2   Sentry                          0x102076528         sentrycrashcm_machexception_getAPI
3   libsystem_pthread.dylib         0x3689cb2bc         _pthread_body
4   libsystem_pthread.dylib         0x3689cb21c         _pthread_start

Thread 17
0   libsystem_kernel.dylib          0x3688afb74         __workq_kernreturn
1   libsystem_pthread.dylib         0x3689cc134         _pthread_wqthread

Thread 18
0   libsystem_kernel.dylib          0x3688afb74         __workq_kernreturn
1   libsystem_pthread.dylib         0x3689cc134         _pthread_wqthread

Thread 19
0   libsystem_kernel.dylib          0x3688afb74         __workq_kernreturn
1   libsystem_pthread.dylib         0x3689cc134         _pthread_wqthread

Thread 20
0   libsystem_kernel.dylib          0x3688afb74         __workq_kernreturn
1   libsystem_pthread.dylib         0x3689cc134         _pthread_wqthread

Thread 21
0   libsystem_kernel.dylib          0x3688afb74         __workq_kernreturn
1   libsystem_pthread.dylib         0x3689cc134         _pthread_wqthread

Thread 22
0   libsystem_kernel.dylib          0x3688afb74         __workq_kernreturn
1   libsystem_pthread.dylib         0x3689cc134         _pthread_wqthread

Thread 23
0   libsystem_kernel.dylib          0x3688a4148         semaphore_timedwait_trap
1   libdispatch.dylib               0x368558884         _dispatch_sema4_timedwait$VARIANT$mp
2   libdispatch.dylib               0x3685591d8         _dispatch_semaphore_wait_slow
3   libdispatch.dylib               0x368566df4         _dispatch_worker_thread
4   libsystem_pthread.dylib         0x3689cb2bc         _pthread_body
5   libsystem_pthread.dylib         0x3689cb21c         _pthread_start



EOF

```

Getting this stack trace for Crashes releated to SmartLook

**Comment by pavel-kroh-smartlook** [comment]
Hi @mithilesh-parmar ,

thank you for the detailed report. 

The crashlog is puzzling, because of the `sl_swizzle_selector_macro_template` is a dead code - it is never ever called. I could be removed from the code and the SDK still compiles and run.

It seems that the symbolication of the crash report is wrong.

We will nevertheless look into the issues and try to guess what might happened, Smartlook uses the DisplayLink, so there is a starting point.

**Comment by js-supp** [comment]
Attaching crash logs received to support.

[591d31fcf72b413d8fce4d3cee2aaa4e-symbolicated.txt](https://github.com/smartlook/smartlook-mobile-issue-tracker/files/7813758/591d31fcf72b413d8fce4d3cee2aaa4e-symbolicated.txt)
[c3dd21de303347efb4c7737fc1361f9a-symbolicated.txt](https://github.com/smartlook/smartlook-mobile-issue-tracker/files/7813760/c3dd21de303347efb4c7737fc1361f9a-symbolicated.txt)

**Comment by mithilesh-parmar** [comment]
Any update on this issue?

**Comment by pavel-kroh-smartlook** [comment]
Hi, still working on it, we expect a maintenances release in a week.

**Comment by pavel-kroh-smartlook** [comment]
Hi @mithilesh-parmar ,

new SDK version has been released. 

It is a maintenance release that also aims to solve some of the crashed. Given the fact that there are ongoing issues with the crashes symbolication, we cannot guarantee we addressed all the crashes, however, we identified some patterns in them and fixed their causes.

New release version is 1.8.4, and it is available via [Swift Package Manager](https://github.com/smartlook/SmartlookSwiftPackage), Cocoapods (`pod Smartlook`) and for [direct download](https://sdk.smartlook.com/ios/smartlook-ios-sdk-1.8.4.52.zip).
